### ISSUE TITLE
DAWN-688 ⁃ Support inline context_free_actions and reading implied actions (was: Events)

### ISSUE THREAD

**Issue by blockone-syncclient** [EOSIO/eos#1754]
After discussing with [~dan.larimer], it was decided that instead of a new first class feature to support structured events consumed by external applications and/or later in-transaction actions we would extend existing features in minor ways to allow the same effects.

1. support sending inline `context_free_action` from webAssembly in parallel with inline actions.  When these actions are dispatched their context-free'ness should be enforced/validated.  Violations should fail the whole transaction
  * This, in conjunction with existing inline-actions, gives a contract author the ability to "send" structure data in a way that is provable to light clients and inter-blockchain-communication.
  * This unifies the way "indexers" would sniff for relevant actions, ie scrape for data sent to a certain receiver/code/action 3-tuple.
  * In cases where there is no on-chain use for the event, context-free-actions can be dispatched to empty or non-existent handlers and provide the same level functionality in a way that is prunable during replay. 
  * Context-free inline actions can be trivially parallelized and dispatched to worker threads ({color:red}NOTE{color}: this is not to be done in this task, but is part of the description of the overall feature)
  * Context-free inline actions should support the notion of context-free-data though this data should not need to persist in any block or transaction structure other than the ephemeral `action_trace`

1. Support reading inline actions generated as part of this transaction, prior to the current executing actions as part of the `get_action` intrinsic.  Inline (non-)context-free actions should be logically appended to their respective lists from the point of view of the calling code.
  * This allows communication of these events forward in the transaction processing.
  * There will be no explicit or convenient utility for addressing these other than index
  * We may need an intrinsic that lets the contract author know the number of available actions of each type(?)

ATC: demonstrable test cases showing that inline actions (context-free or otherwise) can be seen in transaction traces simulating an external observer AND in future actions in the transaction.

**Comment by blockone-syncclient** [comment]
➤ Bart Wyatt commented:

I would like review from [~areg.hayrapetian] and [~kevin.heifner] at least.  [~greg.lee] also needs to be made aware of this as the resolution to one of the outstanding "maybe 3.0" features from the whiteboard session on Monday 3/19/18

It is possible to move this out of 3.0 however, Dan would like efforts made to get it in as it will be a hardfork.

**Comment by bytemaster** [comment]
Created a new issue for the API to read the inline actions, API can be added later, we already support generating the inline actions. 

https://github.com/EOSIO/eos/issues/1843

**Comment by andriantolie** [comment]
Related PR is #1866

**Comment by andriantolie** [comment]
ATC passes.

This issue is already covered by api_tests/cf_action_tests (read_inline_action and read_inline_cf_action) and the unit test passes.
